### DISCUSSION TITLE
Custom binary type is ignored, got constructed with external_constructor<value_t::array> instead

### DISCUSSION THREAD

**Discussion by lemonpy** [nlohmann/json#4200] [Q&A]
The binary representation I want to use is `std::vector<char>`  but the following test fails

```
using json = nlohmann::basic_json<
    std::map,
    std::vector,
    std::string,
    bool,
    std::int64_t,
    std::uint64_t,
    double,
    std::allocator,
    nlohmann::adl_serializer,
    std::vector<char> /* binary_t */>;

TEST_CASE("copy assign binary_t without explicit ctor")
{
    json empty;
    std::vector<char> buf(5, '1');
    empty["buf"] = buf;
    REQUIRE(empty["buf"].is_binary());
}
```

Instead of using the `external_constructor<value_t::binary>` I get the `external_constructor<value_t::array>` .
I tried adding something like `!std::is_same<typename BasicJsonType::binary_t::container_type, CompatibleArrayType>::value` in the template arguments and works for my case but 
```
===============================================================================
/home/lemon/repos/json_fork/tests/src/unit-bjdata.cpp:208:
TEST CASE:  BJData
  individual values
  binary
  Other Serializations
  No Count No Type

~/repos/json_fork/tests/src/unit-bjdata.cpp:1731: ERROR: CHECK( json::from_bjdata(result) == j_out ) is NOT correct!
  values: CHECK( [120,120,120,120,120,120,120,120,120,120] == {"bytes":[120,120,120,120,120,120,120,120,120,120],"subtype":null} )

~/repos/json_fork/tests/src/unit-bjdata.cpp:1732: ERROR: CHECK( json::from_bjdata(result, true, false) == j_out ) is NOT correct!
  values: CHECK( [120,120,120,120,120,120,120,120,120,120] == {"bytes":[120,120,120,120,120,120,120,120,120,120],"subtype":null} )

===============================================================================
~/repos/json_fork/tests/src/unit-bjdata.cpp:208:
TEST CASE:  BJData
  individual values
  binary
  Other Serializations
  Yes Count No Type

~/repos/json_fork/tests/src/unit-bjdata.cpp:1758: ERROR: CHECK( json::from_bjdata(result) == j_out ) is NOT correct!
  values: CHECK( [120,120,120,120,120,120,120,120,120,120] == {"bytes":[120,120,120,120,120,120,120,120,120,120],"subtype":null} )

~/repos/json_fork/tests/src/unit-bjdata.cpp:1759: ERROR: CHECK( json::from_bjdata(result, true, false) == j_out ) is NOT correct!
  values: CHECK( [120,120,120,120,120,120,120,120,120,120] == {"bytes":[120,120,120,120,120,120,120,120,120,120],"subtype":null} )

===============================================================================
[doctest] test cases:     25 |     24 passed |   1 failed | 0 skipped
[doctest] assertions: 692193 | 691653 passed | 540 failed |
[doctest] Status: FAILURE!
```
So this BJData is an array by design? Could it be a binary_type_with_subtype?
Thanks

**Comment by nlohmann** [comment]
When you assign a raw std::vector<char> to a json value via operator=, the library's type traits see it first as a CompatibleArrayType (any container matching your ArrayType) rather than as your custom binary_t, because binary_t itself is just an alias for the same std::vector<char>, and the array-conversion overload is picked before an explicit binary interpretation is considered. To force binary construction, wrap the value explicitly, e.g. empty["buf"] = json::binary(buf) (or basic_json(binary_t(buf))), instead of relying on implicit conversion. The same applies to your BJData round-trip: without an explicit binary_t construction, uint8 arrays stay tagged as JSON arrays, so from_bjdata correctly reconstructs them as arrays, not as binary values — this isn't a bug, it's the expected disambiguation behavior given binary_t and ArrayType overlap for std::vector<char>.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
